File: sim/NationRise.Core/Data/WorldFile.cs
using NationRise.Core.World;

namespace NationRise.Core.Data;

public sealed class WorldFileException(string message) : Exception(message);

/*
   Reads the binary the map pipeline produces. The format carries no geometry:
   the simulation never needs coordinates, and leaving them out makes it
   impossible for rendering concerns to leak in through the data layer.
*/
public static class WorldFile
{
    private const uint Magic = 0x4e525744;
    private const ushort SupportedVersion = 2;

    public static WorldData Read(Stream stream)
    {
        ArgumentNullException.ThrowIfNull(stream);
        using var reader = new BinaryReader(stream, System.Text.Encoding.ASCII, leaveOpen: true);

        uint magic = reader.ReadUInt32();
        if (magic != Magic)
        {
            throw new WorldFileException($"Not a world file: magic 0x{magic:x8}.");
        }

        ushort version = reader.ReadUInt16();
        if (version != SupportedVersion)
        {
            throw new WorldFileException(
                $"World file version {version} is not supported (expected {SupportedVersion}).");
        }

        ushort nationCount = reader.ReadUInt16();
        int provinceCount = checked((int)reader.ReadUInt32());
        _ = reader.ReadUInt32();

        var tags = new string[nationCount];
        for (int i = 0; i < nationCount; i++)
        {
            tags[i] = System.Text.Encoding.ASCII.GetString(reader.ReadBytes(3)).Trim();
        }

        var owner = ReadArray<ushort>(reader, provinceCount, sizeof(ushort), BitConverter.ToUInt16);
        var terrain = reader.ReadBytes(provinceCount);
        var population = ReadArray<float>(reader, provinceCount, sizeof(float), BitConverter.ToSingle);
        var isCity = reader.ReadBytes(provinceCount);
        var resource = reader.ReadBytes(provinceCount);
        var claimOffsets = ReadArray<int>(reader, provinceCount + 1, sizeof(int), BitConverter.ToInt32);
        var claims = ReadArray<ushort>(reader, claimOffsets[provinceCount], sizeof(ushort), BitConverter.ToUInt16);

        var land = ReadGraph(reader, provinceCount);
        var sea = ReadGraph(reader, provinceCount);

        return new WorldData(tags, owner, terrain, population, isCity, resource, claimOffsets, claims, land, sea);
    }

    private static ProvinceGraph ReadGraph(BinaryReader reader, int provinceCount)
    {
        int edgeCount = checked((int)reader.ReadUInt32());
        var offsets = ReadArray<int>(reader, provinceCount + 1, sizeof(int), BitConverter.ToInt32);
        var neighbours = ReadArray<ushort>(reader, edgeCount, sizeof(ushort), BitConverter.ToUInt16);
        return new ProvinceGraph(offsets, neighbours);
    }

    private static T[] ReadArray<T>(BinaryReader reader, int count, int size, Func<byte[], int, T> convert)
    {
        byte[] bytes = reader.ReadBytes(count * size);
        if (bytes.Length != count * size)
        {
            throw new WorldFileException($"Truncated file: expected {count * size} bytes, read {bytes.Length}.");
        }

        var result = new T[count];
        for (int i = 0; i < count; i++)
        {
            result[i] = convert(bytes, i * size);
        }

        return result;
    }
}

public sealed class ProvinceGraph(int[] offsets, ushort[] neighbours)
{
    public ReadOnlySpan<ushort> NeighboursOf(int province)
    {
        int start = offsets[province];
        return neighbours.AsSpan(start, offsets[province + 1] - start);
    }

    public int EdgeCount => neighbours.Length / 2;
}

public sealed class WorldData(
    string[] nationTags,
    ushort[] owner,
    byte[] terrain,
    float[] population,
    byte[] isCity,
    byte[] resource,
    int[] claimOffsets,
    ushort[] claims,
    ProvinceGraph land,
    ProvinceGraph sea)
{
    public string[] NationTags { get; } = nationTags;
    public int ProvinceCount { get; } = owner.Length;
    public ProvinceGraph Land { get; } = land;
    public ProvinceGraph Sea { get; } = sea;

    public Economy.Resource ResourceOf(int province) => (Economy.Resource)resource[province];

    public WorldState ToWorldState(ulong seed)
    {
        var provinces = new ProvinceStore(ProvinceCount, claimOffsets, claims);
        var nations = new NationStore(NationTags.Length);

        for (int i = 0; i < NationTags.Length; i++)
        {
            nations.Tag[i] = NationTags[i];
            nations.Name[i] = NationTags[i];
        }

        for (int i = 0; i < ProvinceCount; i++)
        {
            provinces.Owner[i] = owner[i];
            provinces.Controller[i] = owner[i];
            provinces.Terrain[i] = terrain[i];
            provinces.Population[i] = population[i];
            provinces.IsCity[i] = isCity[i] != 0;
            provinces.Morale[i] = 0.70f;
        }

        return new WorldState(
            provinces,
            nations,
            new Time.GameClock(),
            new Determinism.DeterministicRandom(seed));
    }
}

File: web/src/sim/data/WorldFile.ts
import { DeterministicRandom } from '~/sim/determinism/DeterministicRandom'
import { Resource } from '~/sim/economy/Resource'
import { GameClock } from '~/sim/time/GameClock'
import { NationStore } from '~/sim/world/NationStore'
import { ProvinceStore } from '~/sim/world/ProvinceStore'
import { WorldState } from '~/sim/world/WorldState'

export class WorldFileError extends Error {
  override readonly name = 'WorldFileError'
}

const MAGIC = 0x4e525744
const SUPPORTED_VERSION = 2
const INITIAL_MORALE = 0.7

/*
 * Pembaca biner yang dihasilkan pipeline peta. Formatnya tidak membawa
 * geometri sama sekali: simulasi tidak pernah butuh koordinat, dan
 * meninggalkannya membuat urusan penyajian mustahil menyusup lewat lapisan
 * data.
 *
 * Berbeda dari versi C# yang menerima Stream, di sini masukannya ArrayBuffer.
 * Simulasi harus bisa berjalan di Worker peramban, tempat berkas tidak ada,
 * sehingga I/O ditinggalkan ke pemanggil.
 */
class Cursor {
  #offset = 0

  constructor(private readonly view: DataView) {}

  get offset(): number {
    return this.#offset
  }

  #ensure(bytes: number): void {
    if (this.#offset + bytes > this.view.byteLength) {
      throw new WorldFileError(
        `berkas terpotong: butuh ${bytes} bita pada offset ${this.#offset}, tersisa ${this.view.byteLength - this.#offset}`,
      )
    }
  }

  uint8(): number {
    this.#ensure(1)
    return this.view.getUint8(this.#offset++)
  }

  uint16(): number {
    this.#ensure(2)
    const value = this.view.getUint16(this.#offset, true)
    this.#offset += 2
    return value
  }

  uint32(): number {
    this.#ensure(4)
    const value = this.view.getUint32(this.#offset, true)
    this.#offset += 4
    return value
  }

  /*
   * Larik dibaca elemen per elemen lewat DataView, bukan dengan membuat
   * pandangan typed array di atas buffer. Offset di dalam berkas ini tidak
   * dijamin selaras: panjang blok tag adalah tiga bita per negara, sehingga
   * larik sesudahnya bisa mulai di alamat ganjil, dan `new Uint16Array(buffer,
   * offset)` melempar untuk offset yang tidak kelipatan dua.
   */
  bytes(count: number): Uint8Array {
    this.#ensure(count)
    const slice = new Uint8Array(this.view.buffer, this.view.byteOffset + this.#offset, count).slice()
    this.#offset += count
    return slice
  }

  uint16Array(count: number): Uint16Array {
    const out = new Uint16Array(count)
    for (let i = 0; i < count; i++) out[i] = this.uint16()
    return out
  }

  int32Array(count: number): Int32Array {
    this.#ensure(count * 4)
    const out = new Int32Array(count)
    for (let i = 0; i < count; i++) {
      out[i] = this.view.getInt32(this.#offset, true)
      this.#offset += 4
    }
    return out
  }

  float32Array(count: number): Float32Array {
    this.#ensure(count * 4)
    const out = new Float32Array(count)
    for (let i = 0; i < count; i++) {
      out[i] = this.view.getFloat32(this.#offset, true)
      this.#offset += 4
    }
    return out
  }

  ascii(count: number): string {
    let out = ''
    for (let i = 0; i < count; i++) out += String.fromCharCode(this.uint8())
    return out.trim()
  }
}

export class ProvinceGraph {
  constructor(
    private readonly offsets: Int32Array,
    private readonly neighbours: Uint16Array,
  ) {}

  neighboursOf(province: number): Uint16Array {
    const start = this.offsets[province]!
    return this.neighbours.subarray(start, this.offsets[province + 1]!)
  }

  get edgeCount(): number {
    return this.neighbours.length / 2
  }
}

export class WorldData {
  readonly provinceCount: number

  constructor(
    readonly nationTags: string[],
    private readonly owner: Uint16Array,
    private readonly terrain: Uint8Array,
    private readonly population: Float32Array,
    private readonly cityFlags: Uint8Array,
    private readonly resource: Uint8Array,
    private readonly claimOffsets: Int32Array,
    private readonly claims: Uint16Array,
    readonly land: ProvinceGraph,
    readonly sea: ProvinceGraph,
  ) {
    this.provinceCount = owner.length
  }

  resourceOf(province: number): Resource {
    return this.resource[province]! as Resource
  }

  toWorldState(seed: bigint | number): WorldState {
    const provinces = new ProvinceStore(this.provinceCount, this.claimOffsets, this.claims)
    const nations = new NationStore(this.nationTags.length)

    for (let i = 0; i < this.nationTags.length; i++) {
      nations.tag[i] = this.nationTags[i]!
      nations.name[i] = this.nationTags[i]!
    }

    for (let i = 0; i < this.provinceCount; i++) {
      provinces.owner[i] = this.owner[i]!
      provinces.controller[i] = this.owner[i]!
      provinces.terrain[i] = this.terrain[i]!
      provinces.population[i] = this.population[i]!
      provinces.isCity[i] = this.cityFlags[i]!
      provinces.morale[i] = INITIAL_MORALE
    }

    return new WorldState(provinces, nations, new GameClock(), new DeterministicRandom(seed))
  }
}

export function readWorld(buffer: ArrayBuffer): WorldData {
  const cursor = new Cursor(new DataView(buffer))

  const magic = cursor.uint32()
  if (magic !== MAGIC) {
    throw new WorldFileError(`bukan berkas dunia: magic 0x${magic.toString(16).padStart(8, '0')}`)
  }

  const version = cursor.uint16()
  if (version !== SUPPORTED_VERSION) {
    throw new WorldFileError(
      `versi berkas dunia ${version} tidak didukung (diharapkan ${SUPPORTED_VERSION})`,
    )
  }

  const nationCount = cursor.uint16()
  const provinceCount = cursor.uint32()
  cursor.uint32() // dilewati; ruang cadangan format

  const tags: string[] = []
  for (let i = 0; i < nationCount; i++) tags.push(cursor.ascii(3))

  const owner = cursor.uint16Array(provinceCount)
  const terrain = cursor.bytes(provinceCount)
  const population = cursor.float32Array(provinceCount)
  const isCity = cursor.bytes(provinceCount)
  const resource = cursor.bytes(provinceCount)
  const claimOffsets = cursor.int32Array(provinceCount + 1)
  const claims = cursor.uint16Array(claimOffsets[provinceCount]!)

  const land = readGraph(cursor, provinceCount)
  const sea = readGraph(cursor, provinceCount)

  return new WorldData(
    tags, owner, terrain, population, isCity, resource, claimOffsets, claims, land, sea,
  )
}

function readGraph(cursor: Cursor, provinceCount: number): ProvinceGraph {
  const edgeCount = cursor.uint32()
  const offsets = cursor.int32Array(provinceCount + 1)
  const neighbours = cursor.uint16Array(edgeCount)
  return new ProvinceGraph(offsets, neighbours)
}
